<!DOCTYPE html>
<html>

<head>
    <meta http-equiv="Content-Type" content="text/html" charset="UTF-8">
    <title>Multi-track RNA-Seq Browser</title>
    <meta name="author" content="Priyank Purohit">
    <link href='http://fonts.googleapis.com/css?family=PT+Sans' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.4.0/css/font-awesome.min.css">
    <link rel="stylesheet" href="http://maxcdn.bootstrapcdn.com/bootstrap/3.3.5/css/bootstrap.min.css">
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.3/jquery.min.js"></script>
    <script src="core/tablesorter.js"></script>
    <link rel="stylesheet" href="core/style.css">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <!-- http://stackoverflow.com/questions/1802936/stop-all-active-ajax-requests-in-jquery -->
    <script type="text/javascript">
    $(function() {
        $.xhrPool = [];
        $.xhrPool.abortAll = function() {
            $(this).each(function(i, jqXHR) { //  cycle through list of recorded connection
                jqXHR.abort(); //  aborts connection
                $.xhrPool.splice(i, 1); //  removes from list by index
            });
        }
        $.ajaxSetup({
            beforeSend: function(jqXHR) {
                $.xhrPool.push(jqXHR);
            }, //  annd connection to list
            complete: function(jqXHR) {
                var i = $.xhrPool.indexOf(jqXHR); //  get index for current connection completed
                if (i > -1) $.xhrPool.splice(i, 1); //  removes from list by index
            }
        });
    })
    </script>
</head>

<body>
    <div class="progress" style="width: 100%; height: 8px; position: fixed; margin-top: 0px; top: 0px; z-index: 100;">
        <div id="progress" class="progress-bar progress-bar-success" role="progressbar" aria-valuenow="1" aria-valuemin="0" aria-valuemax="113" style="width:1%;"></div>
    </div>
    <div class="container">
        <!-- Title + Help + Cite -->
        <div class="row">
            <div class="col-xs-10 col-sm-10 col-md-10">
                <p style="font-size: 25px; margin: 5px; text-align: center;"><i>A. Thaliana</i> Multi-track RNA-Seq Browser</p>
                <p style="font-size: 15px; margin: 5px; text-align: center;">Priyank Purohit, Dr. Nicholas Provart. University of Toronto.</p>
            </div>
            <div class="col-xs-2 col-sm-2 col-md-2" style="padding: 15px;">
                <div class="col-xs-4">
                </div>
                <div class="col-xs-4">
                    <button type="button" class="btn btn-default">Cite</button>
                </div>
                <div class="col-xs-4">
                    <button type="button" class="btn btn-info">Help</button>
                </div>
            </div>
        </div>
        <hr style="border: 1px solid gray; margin: 5px; " />
        <!-- Options and Variants selection. -->
        <div class="row">
            <!-- Variants selection -->
            <div class="col-xs-12 col-sm-12 col-md-6">
                <div id="variants_div">
                    <!-- Prepopulated with AT2G24270 data. -->
                    <div class="radio">
                        <label>
                            <input type="radio" name="radio_group" value="0" checked="checked"><img id="variant_0" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAcIAAAAIBAMAAACYMuIQAAAAFVBMVEX///8AAADcFDz/jAAAAP+m 3KYAfQDytQt7AAAAS0lEQVQ4jWMIxQfSoCCBAQRgvDRUHhYAVsCGWyABhZuATRZFO8wEOEBIpuL1 ABAwhAYOex8KDncfBg57HwoOdzAC4nD458NhX5YCAOtozsHok4ONAAAAAElFTkSuQmCC " class="">
                        </label>
                    </div>
                    <div class="radio">
                        <label>
                            <input type="radio" name="radio_group" value="1"><img id="variant_1" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAcIAAAAIBAMAAACYMuIQAAAAFVBMVEX///8AAADcFDz/jAAAAP+m 3KYAfQDytQt7AAAAT0lEQVQ4jWNgDcUD0qAggQEEYLw0VB4WAFbAhlsgAYWbgE0WRTvMBDhAlkxA EUpF8UAAUII1cNj7UHC4+5Bx2PtQcLiDERCHwz8fDvuyFACDg7uy8+q5lAAAAABJRU5ErkJggg== " class="">
                        </label>
                    </div>
                    <div class="radio">
                        <label>
                            <input type="radio" name="radio_group" value="2"><img id="variant_2" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAcIAAAAIBAMAAACYMuIQAAAAFVBMVEX///8AAADcFDz/jAAAAP+m 3KYAfQDytQt7AAAAUElEQVQ4jWNggIJQbCANChLASmC8NFQeFgBWwIZbIAGFm4BNFkU7zAQ4QEim BkCFWNHcHgBXzDjsfSg43H3IOOx9KDjcwQiIw+GfD4d9WQoAgkyakcEOyvoAAAAASUVORK5CYII=" class="">
                        </label>
                    </div>
                    <div class="radio">
                        <label>
                            <input type="radio" name="radio_group" value="3"><img id="variant_3" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAcIAAAAIBAMAAACYMuIQAAAAFVBMVEX///8AAADcFDz/jAAAAP+m 3KYAfQDytQt7AAAAUElEQVQ4jWNggIJQEAhLS2NABmxpEJAA5qXBACoPC0DRjEUgAYWbgE0WRXsa mruQJFNDAxgIAMZh70PB4e5DxmHvQ8HhDkZAHA7/fDjsy1IA2bJX0qO2690AAAAASUVORK5CYII=" class="">
                        </label>
                    </div>
                </div>
            </div>
            <!-- Options -->
            <div class="col-xs-12 col-sm-12 col-md-6">
                <div class="row">
                    <!-- SVG colouring mode: abs or rel -->
                    <div class="col-xs-4">
                        <p>
                            <div class="radio">
                                <label>
                                    <input type="radio" name="svg_colour_radio_group" value="abs" checked="checked">Absolute
                                </label>
                                <label style="margin-left: 10px;">
                                    <input type="radio" name="svg_colour_radio_group" value="rel">Relative
                                </label>
                            </div>
                            <i>SVG Colouring Mode</i>
                        </p>
                    </div>
                    <div class="col-xs-1">
                        <input id="svg_colour_button" type="button" class="btn" onclick="colour_svgs_now()" value="Go" style="height: 35px;" />
                    </div>
                    <div class="col-xs-1"></div>
                    <!-- Locus input -->
                    <div class="col-xs-4">
                        <p>
                            <input type="text" class="form-control" name="Locus" id="locus" value="AT2G24270" style="width: 175px; height: 35px;" /><i>Locus</i>
                        </p>
                    </div>
                    <div class="col-xs-1">
                        <input id="locus_button" type="button" class="btn btn-success" onclick="update_all_images(0)" value="Go" style="height: 35px;" />
                    </div>
                </div>
                <div class="row">
                    <!-- Y axis scaling -->
                    <div class="col-xs-4">
                        <p>
                            <input type="text" class="form-control" name="yscale_inp" id="yscale_input" value="Auto" style="width: 175px; height: 35px;" /><i>RNA-Seq Y-axis Scale</i>
                        </p>
                    </div>
                    <div class="col-xs-1">
                        <input id="yscale_button" type="button" class="btn" onclick="" value="Go" style="height: 35px;" />
                    </div>
                    <div class="col-xs-1"></div>
                    <!-- Absolute SVG scaling -->
                    <div class="col-xs-4">
                        <p>
                            <input type="text" class="form-control" name="rpkm_inp" id="rpkm_scale_input" value="1000" style="width: 175px; height: 35px;" /><i>eFP-RPKM Abs Scale</i>
                        </p>
                    </div>
                    <div class="col-xs-1">
                        <input id="abs_scale_button" type="button" class="btn" onclick="colour_svgs_now('abs')" value="Go" style="height: 35px;" />
                    </div>
                </div>
            </div>
        </div>
        <hr style="border: 1px solid gray; margin: 5px; " />
        <!-- RNA-Seq table. -->
        <div class="col-xs-12 table-responsive">
            <table id="thetable" class="table"></table>
        </div>
    </div>
    <script type="text/javascript">
    // Get initial values
    var colouring_mode = $('input[type="radio"][name="svg_colour_radio_group"]:checked').val();
    var locus = document.getElementById("locus").value; // Which locus does the user want?
    var yscale_input = document.getElementById("yscale_input").value; // What Y-Scale does the user want?
    var max_abs_scale = document.getElementById("rpkm_scale_input").value; // What is the max value for the svg colouring in absolute mode?

    // Locus information
    var locus_start = 10326918; // Gets updated when user changes gene
    var locus_end = 10330048; // Gets updated when user changes gene
    var splice_variants = '';

    // Locus-RNASeq specific information
    var rnaseq_calls = []; // Make a list of records and tissues we need to query....
    var exp_info = []; // Keep track of the FPKM related information // TODO : rename all exp_info to fpkm_info
    var rnaseq_success = 0; // Make a list of records and tissues we need to query....
    var max_absolute_fpkm = -1;
    var max_log_fpkm = -1;

    // Base 64 images
    var img_loading_base64 = "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAcIAAAAyCAYAAADP/dvoAAAABGdBTUEAALGPC/xhBQAAAAlwSFlzAAAOwgAADsIBFShKgAAAABh0RVh0U29mdHdhcmUAcGFpbnQubmV0IDQuMC45bDN+TgAAA7NJREFUeF7t26FTFk8cB2ADgT+AQDAYCQQDgWAgGAwEI4FgIBIMBoMzBAKBQCQYCASDgUAgEAhEAoFAIBgMBIPBYCDcj8/+3sV3Xk/kdTTAPs/Md969vb27dPOZ3dv3UQcADROEADRNEALQtJ+D8NF1l1JKKfVQa0RPz8gFSiml1EOqET09/QMB4F4ThAA0TRAC0DRBCEDTBCEATROEADRNEALQNEEIQNMEIQBNE4QANE0QAtA0QQhA0wQhAE0ThAA0TRDC+La3t7vLy8vB0d81OTk5aHXd69ev/9lzgAFBCON78uRJd3R0NDj6ux4N3rOrq6vu1atX3efPn8sx8I8IQhjfr4Iw4XV6etodHx93X79+HfT+78uXL+Was7OzQc8P5+fn3cnJSbm+BmHUvsiYuLi4KM8YlfO5fyrXAXckCGF8fUGY4Hv27Fn39OnT7uXLl93U1NTNmN3d3dKfGd7s7Gz5rdbW1rrHjx93z58/LzUchGl/+vSptPPMN2/edAsLC6U9fI+lpaXuxYsX5fqJiYlueXl5cAb4rV/kW09P/0BoUV8Qrq+vlwCs8h1xfn6+tL99+3Yzs0tg1rBLyOWbYF3+/Pjx4825GA3Czc3N0s74nMv3w1RCt94/Qbu/v1/awB3knRt676qenv6B0KK+IEwIZuZXJcCGQ+3w8LDM/lZWVm76c4/ca9htQVjbUc9lyTVhmqXR79+/dzMzM5ZGYRx554beu6qnp38gtKgvCBcXF7v3798Pjn7M9mJjY6MsmyYMhwPybwRhZpgJv7m5uTIbzMwUGEPeuaH3rurp6R8ILeoLwgRQwq4uUWYZsy6V5jdhGJm51bDLsma+6dUNNFtbWzfn4i5BmI05CcCDg4PSHv67Rd2EE5ktfvjwobQjoWxHKlzLOzf03lU9Pf0DoUWZgSWIaiX0MjPLhpVsfMn5bI6pwbW3t1e+42WjS8akXSX8pqeny4xudXW1XF/dFoQJ0BznubkmG2SyaSbtt2/fljHZXJPnRQIx98tSaiQ86zdHaNr1e1FqhCCEP5SNMQmnUQmgGkKjMlvLdX8iM80EaJXAG11uBW4hCOF+yxJtZpiZ/WUzTmaW7969G5wFfksQwv2XP9nv7OyUqn+8B+5IEALQNEEIQNMEIQBNE4QANE0QAtA0QQhA0wQhAE0ThAA0TRAC0DRBCEDTBCEATROEADRNEALQNEEIQNMEIQBNGzsIlVJKqYdYI3p6Ri5QSimlHlKN+LkHABoiCAFoWNf9Bx2Q2ZDpi98WAAAAAElFTkSuQmCC";
    var img_gene_struct_1 = "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAcIAAAAIBAMAAACYMuIQAAAAFVBMVEX///8AAADcFDz/jAAAAP+m 3KYAfQDytQt7AAAAS0lEQVQ4jWMIxQfSoCCBAQRgvDRUHhYAVsCGWyABhZuATRZFO8wEOEBIpuL1 ABAwhAYOex8KDncfBg57HwoOdzAC4nD458NhX5YCAOtozsHok4ONAAAAAElFTkSuQmCC ";

    /* Produces an intermediate HEX colour. */
    function generate_colour(start_color, end_color, percent) {
        // strip the leading # if it's there
        start_color = start_color.replace(/^\s*#|\s*$/g, '');
        end_color = end_color.replace(/^\s*#|\s*$/g, '');
        // convert 3 char codes --> 6, e.g. `E0F` --> `EE00FF`
        if (start_color.length == 3) {
            start_color = start_color.replace(/(.)/g, '$1$1');
        }
        if (end_color.length == 3) {
            end_color = end_color.replace(/(.)/g, '$1$1');
        }
        // get colors
        var start_red = parseInt(start_color.substr(0, 2), 16),
            start_green = parseInt(start_color.substr(2, 2), 16),
            start_blue = parseInt(start_color.substr(4, 2), 16);

        var end_red = parseInt(end_color.substr(0, 2), 16),
            end_green = parseInt(end_color.substr(2, 2), 16),
            end_blue = parseInt(end_color.substr(4, 2), 16);
        // calculate new color
        var diff_red = end_red - start_red;
        var diff_green = end_green - start_green;
        var diff_blue = end_blue - start_blue;
        diff_red = ((diff_red * percent) + start_red).toString(16).split('.')[0];
        diff_green = ((diff_green * percent) + start_green).toString(16).split('.')[0];
        diff_blue = ((diff_blue * percent) + start_blue).toString(16).split('.')[0];
        // ensure 2 digits by color
        if (diff_red.length == 1)
            diff_red = '0' + diff_red
        if (diff_green.length == 1)
            diff_green = '0' + diff_green
        if (diff_blue.length == 1)
            diff_blue = '0' + diff_blue
        return '#' + diff_red + diff_green + diff_blue;
    };

    /* Find and colour SVG in the DOM. */
    function colour_part_by_id(id, part, fpkm, mode) {
        // Get the user set RPKM scale
        max_abs_scale = document.getElementById("rpkm_scale_input").value;
        if ((!max_abs_scale) || max_abs_scale <= 0)
            max_abs_scale = 1000;

        if (document != null) {
            var element = document.getElementById(id);
            if (element != null) {
                var SVG = element.getSVGDocument();
                if (SVG != null) {
                    var part = SVG.getElementById("Shapes");
                    if (part != null) {
                        var paths = part.getElementsByTagName("path");
                        if (paths != null) {
                            if (mode == "abs") { // For absolute FPKM colouring
                                var r = 255;
                                var g = 255 - parseInt(fpkm / max_abs_scale * 255);
                                var b = 0;
                                //console.log('abs fpkm = ' + fpkm + ' -> rgb(' + r + ', ' + g + ', ' + b + ')');
                                for (i = 0; i < paths.length; i++) {
                                    paths[i].style.fill = 'rgb(' + r + ', ' + g + ', ' + b + ')';
                                }
                            } else if (mode == "rel") { // For relative FPKM colouring
                                var hex = "";
                                // Make the log FPKM a number between 0 and 1 to denote the 0 to +-5 scale.
                                var log_scaling = 0;
                                if (Math.abs(fpkm) > 5)
                                    log_scaling = 5;
                                else
                                    log_scaling = Math.abs(fpkm);
                                log_scaling /= 5;

                                if (fpkm > 0) { // yellow-red
                                    hex = generate_colour("FFFF00", "FF0000", log_scaling);
                                } else if (fpkm == 0) { // yellow
                                    hex = "FFFF00";
                                } else if (fpkm < 0) { // yellow-blue
                                    hex = generate_colour("FFFF00", "0000FF", log_scaling);
                                }
                                //console.log('fpkm = ' + fpkm + ' -> hex = ' + hex);
                                for (i = 0; i < paths.length; i++) {
                                    paths[i].style.fill = hex;
                                }
                            }
                            document.getElementById(id.replace('_svg', '_rpkm')).innerHTML = fpkm;
                        } else {
                            console.log("Paths is null");
                        }
                    } else {
                        console.log("Part is null");
                    }
                } else {
                    console.log("SVG is null");
                }
            } else {
                console.log("Element is null");
            }
        } else {
            console.log("Document is null");
        }
    }

    /* Finds and updates each SVG in the DOM. */
    function colour_svgs_now(mode = $('input[type="radio"][name="svg_colour_radio_group"]:checked').val()) {
        //console.log("colour_svgs_now function is called with mode = " + mode);
        for (var i = 0; i < 113; i++) {
            // For every exp, figure out the fpkm average of the controls
            var ctrl_fpkm_sum = 0;
            var ctrl_count = 0;
            var ctrl_avg_fpkm = 0;
            for (var ii = 0; ii < 113; ii++) {
                if (exp_info[i][2].indexOf(exp_info[ii][0].slice(0, -4)) != -1) {
                    // experiment ii is a control for experiment i, save FPKM of exp ii
                    ctrl_count++;
                    ctrl_fpkm_sum += exp_info[ii][3];
                }
            }
            if (ctrl_count > 0)
                ctrl_avg_fpkm = ctrl_fpkm_sum / ctrl_count;

            // Save the average fpkm of controls and the log fpkm...
            exp_info[i].splice(4, 1, Math.log2(exp_info[i][3] / ctrl_avg_fpkm));
            exp_info[i].splice(6, 1, ctrl_avg_fpkm);

            // See if the absolute or the relative FPKM is max
            if (exp_info[i][3] >= max_absolute_fpkm)
                max_absolute_fpkm = exp_info[i][3];
            if (Math.abs(exp_info[i][4]) >= max_log_fpkm && Math.abs(exp_info[i][4]) < 1000)
                max_log_fpkm = Math.abs(exp_info[i][4]);

            // Colour SVGs based on the mode requested. Pass in the correct FPKM value...
            if (mode == "rel") {
                if (!exp_info[i][4])
                    exp_info[i][4] = -1;
                colour_part_by_id(exp_info[i][0], exp_info[i][1], exp_info[i][4], mode); // index 5 = relative fpkm
            } else {
                colour_part_by_id(exp_info[i][0], exp_info[i][1], exp_info[i][3], mode); // index 3 = absolute fpkm
            }
        }
        //console.log('Max ABS FPKM = ' + max_absolute_fpkm);
        //console.log('Max Log FPKM = ' + max_log_fpkm);
        //console.log("Colouring function finished. Errors should be above this log entry.");

        $('#thetable').tablesorter();
        $("#thetable").trigger("update");
    }

    /* Re-read the value from the input box*/
    function get_input_values() {
        colouring_mode = $('input[type="radio"][name="svg_colour_radio_group"]:checked').val();
        locus = document.getElementById("locus").value;
        yscale_input = document.getElementById("yscale_input").value;
        max_abs_scale = document.getElementById("rpkm_scale_input").value;
    }

    /* When user clicks GO button, this function is called to update gene structure AND RNA-Seq images*/
    function update_all_images(status) {
        $.xhrPool.abortAll();
        variants_radio_options(status);
    }

    /* Updates the radio button <DIV> with new variants images. */
    function variants_radio_options(status) {
        get_input_values();
        $.ajax({
            url: 'http://bar.utoronto.ca/~ppurohit/RNA-Browser/cgi-bin/get_gene_structures.cgi?locus=' + locus,
            dataType: 'json',
            success: function(gene_res) {
                // Update locus_start and locus_end
                locus_start = gene_res['locus_start'];
                locus_end = gene_res['locus_end'];
                splice_variants = JSON.stringify(gene_res['splice_variants']);
                populate_table(status);

                // Remove existing variant images.
                var variants_div = document.getElementById("variants_div");
                while (variants_div.firstChild) {
                    variants_div.removeChild(variants_div.firstChild);
                }
                for (var i = 0; i < parseInt(gene_res['variant_count']); i++) {
                    // retrieve the base64 and create the element to insert
                    var append_str = "<div class=\"radio\"><label><input type=\"radio\" value=\"" + i + "\"";
                    if (i == 0) { // The first gene structure should be selected by default
                        append_str += "checked=\"checked\"";
                    }
                    append_str += " name=\"radio_group\"><img id=\"variant_" + i + "\" src=\"data:image/png;base64," + gene_res['splice_variants'][i]['gene_structure'] + "\" ></label></div>";
                    // Append the element to the div
                    $("#variants_div").append(append_str);
                }
                img_gene_struct_1 = "data:image/png;base64," + gene_res['splice_variants'][0]['gene_structure'];
                var all_gene_structure_imgs = document.getElementsByClassName('gene_structure_img');
                for (var i = 0; i < all_gene_structure_imgs.length; i++) {
                    all_gene_structure_imgs[i].src = "data:image/png;base64," + gene_res['splice_variants'][0]['gene_structure'];
                }
                $('input[type=radio][name=radio_group]').change(function() { // Bind an event listener..
                    gene_structure_radio_on_change();
                });
            }
        });
    }

    /* When radio button changes, update the gene structure throughout the document and update the PCC values */
    function gene_structure_radio_on_change() {
        // Find out which variant is currently selected
        var variant_selected = $('input[type="radio"][name="radio_group"]:checked').val();
        // Get the variant's img src
        var base64 = document.getElementById('variant_' + variant_selected).src;
        // Find all img tags that should be updated (all the <img> with class gene_structure)
        var all_gene_structure_imgs = document.getElementsByClassName('gene_structure_img');
        // Change their src to the newly selected variant's src
        for (var i = 0; i < all_gene_structure_imgs.length; i++) {
            all_gene_structure_imgs[i].src = base64;
        }
        // update all pcc pcc_value
        // Go through the exp_info array and make changes
        for (var i = 0; i < exp_info.length; i++) {
            //console.log("exp_info[i] = " + exp_info[i]);
            document.getElementById(rnaseq_calls[i][1] + '_pcc').innerHTML = exp_info[i][5][variant_selected];
        }
    }

    function parseIntArray(arr) {
        for (var i = 0, len = arr.length; i < len; i++) {
            arr[i] = parseInt(arr[i], 10);
        }
        return arr;
    }

    /* Makes AJAX request for each RNA-Seq image based on the rnaseq_calls array that was produced by the populate_table() function */
    function rnaseq_images(status) {
        rnaseq_success = 0;
        get_input_values();
        if (rnaseq_calls.length == 113) {
            for (var i = 0; i < 113; i++) {
                $.ajax({
                    url: 'http://ec2-52-70-232-122.compute-1.amazonaws.com/~ppurohit/RNA-Browser/cgi-bin/webservice.cgi?tissue=' + rnaseq_calls[i][0] + '&record=' + rnaseq_calls[i][1] + '&locus=' + locus + '&variant=1&start=' + locus_start + '&end=' + locus_end + '&status=' + status + '&struct=' + splice_variants,
                    dataType: 'json',
                    success: function(response_rnaseq) {
                        if (locus != response_rnaseq['locus']) {
                            console.log("ERROR: " + locus + "'s RNA-Seq API request returned with data for some other locus.");
                        }
                        // Update the progress bar
                        if (response_rnaseq['status'] == 200) {
                            rnaseq_success++;
                            $('div#progress').width(rnaseq_success / 113 * 100 + '%');
                        } else {
                            console.log("ERROR CODE = " + response_rnaseq['status'] + " returned for " + locus + " RNA-Seq data.");
                        }
                        var r = [];
                        if (status == 1) {
                            // Finalize statistical calculations
                            var ss_y = parseInt(response_rnaseq['ss_y']);
                            var sum_y = parseInt(response_rnaseq['sum_y']);
                            var ssy = parseInt(response_rnaseq['ss_y']);
                            var sum_xy = parseIntArray(response_rnaseq['sum_xy'].replace(/\[/g, "").replace(/\]/g, "").replace(/"/g, "").split(','));
                            var sum_x = parseIntArray(response_rnaseq['sum_x'].replace(/\[/g, "").replace(/\]/g, "").replace(/"/g, "").split(','));
                            var sum_xx = parseIntArray(response_rnaseq['sum_xx'].replace(/\[/g, "").replace(/\]/g, "").replace(/"/g, "").split(','));
                            var ss_x = parseIntArray(response_rnaseq['ss_x'].replace(/\[/g, "").replace(/\]/g, "").replace(/"/g, "").split(','));
                            var ssx = parseIntArray(response_rnaseq['ss_x'].replace(/\[/g, "").replace(/\]/g, "").replace(/"/g, "").split(','));
                            var n = parseInt(response_rnaseq['end']) - parseInt(response_rnaseq['start']);
                            var sp = [];
                            // Compute the r values for each variant
                            for (var i = 0; i < sum_xy.length; i++) {
                                sp.splice(i, 0, sum_xy[i] - ((sum_x[i] * sum_y) / n));
                                r.splice(i, 0, sp[i] / (Math.sqrt(ssx[i] * ssy)));
                            }
                        } else {
                            //r.push(parseIntArray(String(response_rnaseq['r']).replace(/\[/g, "").replace(/\]/g, "").replace(/"/g, "").split(',')));
                            r = response_rnaseq['r'];
                            console.log(r);
                        }
                        //console.log("ss_y = ", ss_y, ", sum_y = ", sum_y, ", sum_xy = ", sum_xy, ", sum_x = ", sum_x, ", sum_xx = ", sum_xx, ", ss_x = ", ss_x, ", ssx = ", ssx, ", ssy = ", ssy, ", n = ", n, ", sp = ", sp, ", ssx = ", ssx, ", ssy = ", ssy, ", r = ", r);

                        // For cache purposes
                        //console.log("if (record == \"" + response_rnaseq['record'] + "\"):");
                        //console.log("\tdumpJSON(" + response_rnaseq['status'] + ", \"" + response_rnaseq['locus'] + "\", " + response_rnaseq['variant'] + ", " + response_rnaseq['chromosome'] + ", " + response_rnaseq['start'] + ", " + response_rnaseq['end'] + ", \"" + response_rnaseq['record'] + "\", \"" + response_rnaseq['tissue'] + "\", \"" + response_rnaseq['rnaseqbase64'] + "\", " + response_rnaseq['reads_mapped_to_locus'] + ", " + response_rnaseq['absolute-fpkm'] + ", " + response_rnaseq['ss_y'] + ", " + response_rnaseq['sum_y'] + ", \"" + response_rnaseq['sum_xy'] + "\", \"" + response_rnaseq['sum_x'] + "\", \"" + response_rnaseq['sum_xx'] + "\", \"" + response_rnaseq['ss_x'] + "\")");

                        //console.log("\tdumpJSON(" + response_rnaseq['status'] + ", \"" + response_rnaseq['locus'] + "\", " + response_rnaseq['variant'] + ", " + response_rnaseq['chromosome'] + ", " + response_rnaseq['start'] + ", " + response_rnaseq['end'] + ", \"" + response_rnaseq['record'] + "\", \"" + response_rnaseq['tissue'] + "\", \"" + response_rnaseq['rnaseqbase64'] + "\", " + response_rnaseq['reads_mapped_to_locus'] + ", " + response_rnaseq['absolute-fpkm'] + ", \"" + response_rnaseq['r'] + "\")");

                        // find the correct row and update coverage image, and stats info
                        document.getElementById(response_rnaseq['record'] + '_rnaseq_img').src = 'data:image/png;base64,' + response_rnaseq['rnaseqbase64'];
                        document.getElementById(response_rnaseq['record'] + '_pcc').innerHTML = r[0];
                        document.getElementById(response_rnaseq['record'] + '_rpkm').innerHTML = response_rnaseq['absolute-fpkm'];

                        // Save the abs-fpkm, and the stats numbers
                        for (var ii = 0; ii < 113; ii++) {
                            if (exp_info[ii][0] == response_rnaseq['record'] + '_svg') { // Find the correct element
                                exp_info[ii].splice(3, 1, response_rnaseq['absolute-fpkm']);
                                exp_info[ii].splice(5, 1, r);
                                //console.log("Found " + response_rnaseq['record'] + " == " + exp_info[ii][0] + ".");
                            }
                        }

                        // Colour SVG by Absolute RPKM
                        colour_part_by_id(response_rnaseq['record'] + '_svg', 'Shapes', response_rnaseq['absolute-fpkm'], 'abs');
                    }
                });
            }
        }
    }

    /* Gets the BAM locator XML to create + populate the table. Leeps track of all RNA-Seq calls it will have to make. */
    function populate_table(status) {
        // Clear up the table + RNA-Seq Calls
        $("#thetable").empty();
        rnaseq_calls = [];
        exp_info = [];
        rnaseq_success = 0;
        max_absolute_fpkm = -1;
        max_log_fpkm = -1;


        // Insert table headers
        $("#thetable").append('<thead><tr>' +
            '<th class="sortable" style="width: 250px;">Title</th>' +
            '<th class="sortable" style="width: 460px;">RNA-Seq Coverage</th>' +
            '<th class="sortable" style="width: 50px;">PCC/Reads</th>' +
            '<th class="sortable" style="width: 100px;">eFP - RPKM</th>' +
            '<th class="sortable" style="width: 50px;">RPKM</th>' +
            '<th class="sortable" style="width: 275px;">Details</th>' +
            '</tr></thead>' +
            '<tbody></tbody>');

        $.ajax({
            url: 'data/bamdata_amazon_links.xml',
            dataType: 'xml',
            success: function(xml_res) {
                var $title = $(xml_res).find("bam_file");
                $title.each(function() { // Iterate over each subtag inside the <file> tag.
                    // Extract information
                    var title = $(this).attr('title');
                    var description = $(this).attr('desc');
                    var svg = $(this).attr('svgname');
                    var experimentno = $(this).attr('record_number');
                    var url = $(this).attr('publication_url');
                    var publicationid = $(this).attr('publication_link');
                    var numberofreads = $(this).attr('total_reads_mapped');
                    var controls = $(this).find("controls")[0].innerHTML.replace(/<bam_exp>/g, "").replace(/<\/bam_exp>/g, ",").replace(/\n/g, " ").replace(/ /g, "").split(",");
                    var links = "";
                    for (var i = controls.length; i--;) {
                        if (controls[i] != "MEDIAN") {
                            links += '<a href="http://www.ncbi.nlm.nih.gov/Traces/sra/?run=' + controls[i] + '" target="blank">' + controls[i] + '</a> ';
                        } else {
                            links += controls[i];
                        }
                    }
                    var name = $(this).attr('bam_link').split("/");
                    var tissue = $(this).attr('bam_link').split("/")[8];
                    var subunitname = $(this).attr('svg_subunit'); // supposed to be the one that gets coloured

                    // Keep track of what experiments we want to query:
                    // Need: tissue, experimentno in rnaseq_calls ... (also need start, end, and locus)
                    rnaseq_calls.push([tissue, experimentno]);

                    // Construct a table row <tr> element
                    var append_str = '<tr>';
                    // Append title <td>
                    append_str += '<td style="width: 250px; font-size: 12px;">' + title + '</td>\n';
                    // Append RNA-Seq and Gene Structure images (2 imgs) in one <td>
                    append_str += '<td style="width: 460px;">' + '<img id="' + experimentno + '_rnaseq_img" width="450px" height="50px" class="rnaseq_img" src="' + img_loading_base64 + '" /><br/>' + '<img id="' + experimentno + '_gene_structure_img" width="450px" height="8px" class="gene_structure_img" src="' + img_gene_struct_1 + '" />' + '</td>\n';
                    // Append the PCC <td>
                    append_str += '<td id="' + experimentno + '_pcc' + '" class="pcc_value" style="font-size: 10px; width: 50px; ">' + numberofreads + '</td>';
                    // Append the approparite SVG with place holder sorting number in front of it .. all in one <td>
                    append_str += '<td style="width:  75px;">' + '<object id="' + experimentno + '_svg' + '" data="SVGs/' + svg.substring(4) + '" width="75" height="75" type="image/svg+xml"></object>' + '</td>\n';
                    // Append abs/rel RPKM
                    append_str += '<td id="' + experimentno + '_rpkm' + '" style="font-size: 10px; width: 50px; ">1000000</td>';
                    // Append the details <td>  
                    append_str += '<td style="width: 200px; font-size: 12px;">' + description + '<br/>' + '<a href="' + url + '" target="blank">' + 'NCBI SRA' + '</a>; <a href="' + publicationid + '" target="blank">PubLink</a>' + '<br/><a href="javascript:(function(){$(\'#' + url.substring(44) + '\').toggle();})()">More Details</a><div id="' + url.substring(44) + '" style="display:none">Total reads = ' + numberofreads + '.<br/>Controls: ' + links + '</div></td>\n';
                    append_str += '</tr>';

                    // Append the <tr> to the table
                    $("#thetable").append(append_str);

                    exp_info.push([experimentno + '_svg', 'Shapes', controls, 0, 0, 0, 0]);

                    if (rnaseq_calls.length == 113) {
                        rnaseq_images(status);
                    }
                });
                $('#thetable').tablesorter();
                $("#thetable").trigger("update");
            }
        });
    }

    function locus_validation() {
        var loc = document.getElementById("locus").value;
        if (loc.length == 9 &&
            (loc[0] == 'A' || loc[0] == 'a') &&
            (loc[1] == 'T' || loc[1] == 't') &&
            ((loc[2] >= 1 && loc[2] <= 5) || loc[2] == 'C' || loc[2] == 'M' || loc[2] == 'c' || loc[2] == 'm') &&
            (loc[3] == 'G' || loc[3] == 'g') &&
            (loc[4] >= 0 && loc[4] <= 9) &&
            (loc[5] >= 0 && loc[5] <= 9) &&
            (loc[6] >= 0 && loc[6] <= 9) &&
            (loc[7] >= 0 && loc[7] <= 9) &&
            (loc[8] >= 0 && loc[8] <= 9)) {
            $("#locus_button").removeAttr('disabled');
        } else {
            $("#locus_button").prop("disabled", true);
        }
    }

    function yscale_validation() {
        var yscale = parseInt(document.getElementById("yscale_input").value);
        if (yscale > 0) {
            $("#yscale_button").removeAttr('disabled');
        } else {
            $("#yscale_button").prop("disabled", true);
        }
    }

    function rpkm_validation() {
        var rpkmscale = parseInt(document.getElementById("rpkm_scale_input").value);
        if (rpkmscale > 0) {
            $("#abs_scale_button").removeAttr('disabled');
        } else {
            $("#abs_scale_button").prop("disabled", true);
        }
    }

    // Leave this function here, even if unused
    $(document).ready(function() {
        // On load, validate input
        locus_validation();
        yscale_validation();
        rpkm_validation();

        // Bind event listeners...
        $('input[type=radio][name=radio_group]').change(function() {
            gene_structure_radio_on_change();
        });

        $('#locus').keyup(function() {
            locus_validation();
        });

        $('#yscale_input').keyup(function() {
            yscale_validation();
        });

        $('#rpkm_scale_input').keyup(function() {
            rpkm_validation();
        });
        populate_table(1); // status 1 forces rna-seq api to return cached data for fast initial load
    });
    </script>
</body>

</html>
